src/v/transform/logging/record_batcher.cc

```diff
@@ -0,0 +1,160 @@
+        return max_records_bytes() - curr_records_bytes_est();
+    }
+    inline size_t batch_size_bytes_est() const {
+        return curr_records_bytes_est()
+               + model::packed_record_batch_header_size;
+    }
+    inline size_t curr_records_bytes_est() const { return _curr_batch_size; }
+    inline size_t max_records_bytes() const {
+        return _batch_max_bytes - model::packed_record_batch_header_size;
+    }
+
+    void roll_batch() {
+        auto batch = std::exchange(_builder, bb_init()).build();
+        auto diff = static_cast<int64_t>(batch_size_bytes_est())
+                    - static_cast<int64_t>(batch.size_bytes());
+        if (diff < 0) {
+            vlog(
+              tlg_log.debug,
+              "Underestimaged batch size {}B - {}B = {}",
+              batch_size_bytes_est(),
+              batch.size_bytes(),
+              diff);
+        } else {
+            vlog(
+              tlg_log.trace,
+              "Building record batch. Actual size: {} (estimated: {}, est - "
+              "act: "
+              "{})",
+              batch.size_bytes(),
+              batch_size_bytes_est(),
+              diff);
+        }
+        // NOTE(oren): We could drop a batch here if it's too large to save
+        // bandwidth, but this should happen very rarely if ever (estimates are
+        // deliberately conservative). The most important thing is that we roll
+        // the batch builder over to make room for subsequent records.
+        _curr_batch_size = 0;
+        _record_batches.push_back(std::move(batch));
+    }
+
+    size_t estimate_record_size(size_t ksize, size_t vsize) {
+        // NOTE(oren): a size estimate for vint fields on an individual record:
+        //    - key size (known)
+        //    - value size (known)
+        //    - offset delta (pessimize)
+        //    - timestamp delta (always nil)
+        //    - headers size (always nil)
+        //    - record size in bytes (calculate based on sum of above)
+        // NOTE(oren): this seems to overestimate the size of the final batch
+        // by ~1%. That's probably acceptably conservative to avoid exceeding
+        // configured limits without losing too much in terms of maximizing
+        // batch size
+        constexpr size_t base_overhead_est
+          = sizeof(model::record_attributes::type) // attrs
+            + vint::max_length    // offset delta, ultra-conservative
+            + vint::vint_size(0)  // timestamp_delta
+            + vint::vint_size(0); // headers size
+        auto sz = vint::vint_size(static_cast<int64_t>(ksize))   // key size
+                  + ksize                                        //
+                  + vint::vint_size(static_cast<int64_t>(vsize)) // val size
+                  + vsize                                        //
+                  + base_overhead_est;
+        return vint::vint_size(static_cast<int64_t>(sz)) + sz;
+    }
```

Comment:
Not your fault, but we really need to centralize these sorts of calculations.

Can we either:
* Create record objects (even if only temporary) to compute the size then release the key and value into the batch builder (or just hold a list of records until we roll)? I don't know if we want to over index on performance here.
* Create a ticket to move all our batch building stuff into the `model` namespace so it can be shared. Feels weird to have to depend on storage anyways.

IIRC the kafka client just creates record objects and at the last second turns them into batches, I think that would be fine here.

Reply:
Yeah it's pretty awful...would be happy to remove, but I was kind of surprised not to find anything more reusable already existing anywhere.

>kafka client just creates record objects and at the last second turns them into batches

I must have missed that, will check it out

Reply:
Wound up with something like:
- in `append` create a `record_batch` with a single record
- estimate the size of the record therein
- perform all the same checks and move the record's guts into the "real" `batch_builder`

I think this is more or less in the spirit of what the internal kafka client is doing, but I might be misreading.